File: apps/mac/PrudenceKit/Sources/PrudenceUI/Fmt.swift
import Foundation
import PrudenceModels

/// Numbers and dates on their way to a screen, through `Locale`.
///
/// `PrudenceModels.Formatting` is the engine-facing half: it parses the engine's UTC stamps
/// and prints the shapes the CLI prints. This is the reader-facing half: the same values in
/// the reader's own language and calendar. Two of them deliberately do not go through
/// `Locale`, and both are marked: the percentage and the plain integer inside an observation
/// sentence, which have to match `store/observations.py` character for character.
public enum Fmt {

    // MARK: - counts

    /// `1,248`, `1 248`, `1,248` — whatever the locale groups with.
    public static func count(_ value: Int, locale: Locale = Localization.locale) -> String {
        let formatter = NumberFormatter()
        formatter.locale = locale
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? String(value)
    }

    /// The same integer with no grouping at all. Used only inside an observation sentence,
    /// where the engine writes `f"{n}"` and the English output has to equal it.
    public static func plain(_ value: Int) -> String { String(value) }

    /// `1.2M`, `43.1k`, `812`. The unit letters are not translated: they are the same in both
    /// languages the app ships, and the mockups print them the same way.
    public static func tokens(_ value: Int, locale: Locale = Localization.locale) -> String {
        if value >= 1_000_000 {
            return decimal(Double(value) / 1_000_000, places: 1, locale: locale) + "M"
        }
        if value >= 1_000 {
            return decimal(Double(value) / 1_000, places: 1, locale: locale) + "k"
        }
        return count(value, locale: locale)
    }

    /// `4.5`, one decimal, locale-aware separator.
    public static func hours(_ value: Double, locale: Locale = Localization.locale) -> String {
        decimal(value, places: 1, locale: locale)
    }

    public static func decimal(
        _ value: Double, places: Int, locale: Locale = Localization.locale
    ) -> String {
        let formatter = NumberFormatter()
        formatter.locale = locale
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = places
        formatter.maximumFractionDigits = places
        return formatter.string(from: NSNumber(value: value)) ?? String(value)
    }

    // MARK: - shares

    /// `88%`, the way the CLI prints a share, and the way an observation sentence must.
    /// Deliberately not `NumberFormatter.percent`, which would write `88 %` in some locales
    /// and break the word-for-word test.
    public static func percent(_ share: Double?) -> String {
        guard let share else { return Str.commonDash.text }
        return String(format: "%.0f%%", share * 100)
    }

    /// The distance between two shares, in points. Neutral: a gap has no sign and no colour.
    public static func points(_ gap: Double) -> String {
        String(format: "%.0f", abs(gap) * 100)
    }

    /// The same distance as a whole number, for the plural rule to choose on.
    public static func pointsValue(_ gap: Double) -> Int { Int((abs(gap) * 100).rounded()) }

    // MARK: - time

    /// `20 Sep 18:04`, or `9月20日 18:04`, in the reader's own time zone and calendar.
    public static func stamp(
        _ date: Date, locale: Locale = Localization.locale, zone: TimeZone = .current
    ) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = zone
        formatter.setLocalizedDateFormatFromTemplate("d MMM HH:mm")
        return formatter.string(from: date)
    }

    /// `4 minutes ago`, `4 分钟前`, `never`.
    public static func relative(
        _ date: Date?, now: Date = Date(), locale: Locale = Localization.locale
    ) -> String {
        guard let date else { return Str.menuNever.text }
        let seconds = now.timeIntervalSince(date)
        if seconds < 60 { return Str.menuJustNow.text }
        let formatter = RelativeDateTimeFormatter()
        formatter.locale = locale
        formatter.unitsStyle = .full
        return formatter.localizedString(for: date, relativeTo: now)
    }

    /// A `yyyy-MM-dd` day as the reader writes it: `20 Sep 2026`, `2026年9月20日`.
    public static func day(_ day: String, locale: Locale = Localization.locale) -> String {
        guard let date = Formatting.date(day) else { return day }
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = .current
        formatter.setLocalizedDateFormatFromTemplate("d MMM y")
        return formatter.string(from: date)
    }

    /// The axis form: `8 Sep`, `9月8日`. No year, because every axis here spans weeks.
    public static func shortDay(_ day: String, locale: Locale = Localization.locale) -> String {
        guard let date = Formatting.date(day) else { return day }
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = .current
        formatter.setLocalizedDateFormatFromTemplate("d MMM")
        return formatter.string(from: date)
    }

    // MARK: - phrases built from a count

    /// The four that carry a count go through the catalog's plural entries, so English says
    /// "1 session" and "3 sessions" and Chinese says "1 个会话" either way.
    public static func sessions(_ value: Int) -> String { Str.unitSessions.plural(value) }
    public static func commits(_ value: Int) -> String { Str.unitCommits.plural(value) }
    public static func edits(_ value: Int) -> String { Str.unitEdits.plural(value) }
    public static func projects(_ value: Int) -> String { Str.unitProjects.plural(value) }

    /// These two carry a measure rather than a count: `6.2k tokens`, `0.1 hours`. A plural
    /// rule has nothing to choose on once the value has been rounded to `6.2k`, so they stay
    /// plain, which is also how the mockups print them.
    public static func tokenPhrase(_ value: Int) -> String { Str.unitTokens(tokens(value)) }
    public static func hourPhrase(_ value: Double) -> String { Str.unitHours(hours(value)) }

    /// Join phrases the way the language joins a list: `a, b, c` and `a，b，c`.
    public static func list(_ parts: [String]) -> String {
        parts.joined(separator: Str.commonListSeparator.text)
    }

    /// The range picker's three labels and the line under the title.
    ///
    /// `ChartRange` lives in `PrudenceModels`, which knows nothing about the interface
    /// language and should not; its `label` and `describe` stay as the English the CLI-facing
    /// layer uses, and the interface says them its own way here.
    public static func range(_ range: ChartRange) -> String {
        switch range {
        case .eightWeeks: return Str.rangeEightWeeks.text
        case .ninetyDays: return Str.rangeNinetyDays.text
        case .all: return Str.rangeAll.text
        }
    }

    public static func rangeDescription(_ range: ChartRange, now: Date) -> String {
        guard let first = range.firstDay(now: now) else { return Str.rangeEveryDay.text }
        return Str.rangeSince(day(first))
    }

    /// The purpose the reader sees. `unknown` is "other", never "unknown".
    public static func purpose(_ key: String) -> String {
        switch key {
        case "development": return Str.purposeDevelopment.text
        case "research": return Str.purposeResearch.text
        case "debugging": return Str.purposeDebugging.text
        case "conversation": return Str.purposeConversation.text
        case "mixed": return Str.purposeMixed.text
        case "unknown": return Str.purposeUnknown.text
        default: return key
        }
    }

    /// The purpose as it appears **inside an observation sentence**.
    ///
    /// English keeps the engine's own label, because that sentence has to equal
    /// `app_observation.sentence` word for word and the engine writes `labelled unknown`
    /// where the interface elsewhere says "other". Every other language gets the reader's
    /// word, since there is no English string to match there.
    public static func purposeInSentence(_ key: String) -> String {
        Localization.running == .english ? key : purpose(key)
    }
}

File: apps/desktop/src/design/fmt.js
/* Numbers and dates on their way to a screen, through the reader's locale.

   The port of `PrudenceUI/Fmt.swift`, function for function. The engine-facing half (what
   the CLI prints, what a UTC stamp means) is not here and is not this layer's business.

   **Two things deliberately do not go through the locale**, and both are marked where
   they are: the `%.0f%%` share and the plain ungrouped integer inside an observation
   sentence, which have to match `store/observations.py` character for character or the
   word-for-word test against `app_observation.sentence` fails. Everything else does. */

(function (global) {
  "use strict";

  var Str = global.Str;

  function locale() {
    return Str.lang();
  }

  /* --- counts ------------------------------------------------------------------- */

  /** `1,248`, `1 248`: whatever the locale groups with. */
  function count(value) {
    return new Intl.NumberFormat(locale(), { maximumFractionDigits: 0 }).format(value);
  }

  /** The same integer with no grouping at all. Only inside an observation sentence,
      where the engine writes `f"{n}"` and the English output has to equal it. */
  function plain(value) {
    return String(value);
  }

  function decimal(value, places) {
    return new Intl.NumberFormat(locale(), {
      minimumFractionDigits: places,
      maximumFractionDigits: places,
    }).format(value);
  }

  /** `1.2M`, `43.1k`, `812`. The unit letters are not translated: they are the same in
      both languages the app ships, and the mockups print them the same way. */
  function tokens(value) {
    if (value >= 1000000) return decimal(value / 1000000, 1) + "M";
    if (value >= 1000) return decimal(value / 1000, 1) + "k";
    return count(value);
  }

  function hours(value) {
    return decimal(value, 1);
  }

  /* --- shares ------------------------------------------------------------------- */

  /** `88%`, the way the CLI prints a share and the way an observation sentence must.
      Deliberately not `Intl.NumberFormat(style: "percent")`, which writes `88 %` in some
      locales and would break the word-for-word test. */
  function percent(share) {
    if (share === null || share === undefined || Number.isNaN(share)) return Str.t("common.dash");
    return Math.round(share * 100) + "%";
  }

  /** The distance between two shares, in points. Neutral: a gap has no sign. */
  function points(gap) {
    return String(Math.round(Math.abs(gap) * 100));
  }

  function pointsValue(gap) {
    return Math.round(Math.abs(gap) * 100);
  }

  /* --- time --------------------------------------------------------------------- */

  function asDate(value) {
    if (value instanceof Date) return value;
    if (typeof value === "number") return new Date(value);
    if (typeof value !== "string") return null;
    var parsed = new Date(value);
    return Number.isNaN(parsed.getTime()) ? null : parsed;
  }

  /** A `yyyy-MM-dd` day as a local Date at midnight, which is what the engine means by a
      day. `new Date("2026-09-21")` would be UTC midnight and can land on the day before. */
  function fromDay(day) {
    var parts = String(day).slice(0, 10).split("-");
    if (parts.length !== 3) return null;
    var date = new Date(Number(parts[0]), Number(parts[1]) - 1, Number(parts[2]));
    return Number.isNaN(date.getTime()) ? null : date;
  }

  /** `20 Sep 18:04`, `9月20日 18:04`. */
  function stamp(value) {
    var date = asDate(value);
    if (!date) return Str.t("common.dash");
    return new Intl.DateTimeFormat(locale(), {
      day: "numeric",
      month: "short",
      hour: "2-digit",
      minute: "2-digit",
      hour12: false,
    }).format(date);
  }

  /** `4 minutes ago`, `4 分钟前`, `never`, `just now`. */
  function relative(value, now) {
    var date = asDate(value);
    if (!date) return Str.t("menu.never");
    var reference = asDate(now) || new Date();
    var seconds = (reference.getTime() - date.getTime()) / 1000;
    if (seconds < 60) return Str.t("menu.justNow");

    var formatter = new Intl.RelativeTimeFormat(locale(), { numeric: "auto", style: "long" });
    var units = [
      ["year", 31536000],
      ["month", 2592000],
      ["week", 604800],
      ["day", 86400],
      ["hour", 3600],
      ["minute", 60],
    ];
    for (var i = 0; i < units.length; i += 1) {
      var size = units[i][1];
      if (Math.abs(seconds) >= size) {
        // Truncated, not rounded: `RelativeDateTimeFormatter` says "13 hours ago" for
        // thirteen and a half, and a panel that says 14 beside one that says 13 on the
        // same data reads as a bug in one of them.
        return formatter.format(-Math.trunc(seconds / size), units[i][0]);
      }
    }
    return Str.t("menu.justNow");
  }

  /** `20 Sep 2026`, `2026年9月20日`. */
  function day(value) {
    var date = fromDay(value);
    if (!date) return String(value);
    return new Intl.DateTimeFormat(locale(), {
      day: "numeric",
      month: "short",
      year: "numeric",
    }).format(date);
  }

  /** The axis form: `8 Sep`, `9月8日`. No year, because every axis here spans weeks. */
  function shortDay(value) {
    var date = fromDay(value);
    if (!date) return String(value);
    return new Intl.DateTimeFormat(locale(), { day: "numeric", month: "short" }).format(date);
  }

  /** `20 Sep 2026 (yesterday)`. Both halves go through the locale. A day is midnight, so
      "0 hours ago" for something written this morning is noise: the relative half counts
      in days and the absolute half carries the precision. */
  function writtenOn(value, now) {
    var absolute = day(value);
    var date = fromDay(value);
    if (!date) return absolute;
    var reference = asDate(now) || new Date();
    var startOfToday = new Date(
      reference.getFullYear(),
      reference.getMonth(),
      reference.getDate()
    );
    var days = Math.round((startOfToday.getTime() - date.getTime()) / 86400000);
    var text = new Intl.RelativeTimeFormat(locale(), { numeric: "auto", style: "long" }).format(
      -days,
      "day"
    );
    return Str.t("common.dateWithRelative", absolute, text);
  }

  /* --- phrases built from a count ----------------------------------------------- */

  /** The four that carry a count go through the plural entries, so English says
      "1 session" and "3 sessions" and Chinese says "1 个会话" either way. */
  function sessions(value) {
    return Str.plural("unit.sessions", value, count(value));
  }
  function commits(value) {
    return Str.plural("unit.commits", value, count(value));
  }
  function edits(value) {
    return Str.plural("unit.edits", value, count(value));
  }
  function projects(value) {
    return Str.plural("unit.projects", value, count(value));
  }

  /** These two carry a measure rather than a count: `6.2k tokens`, `0.1 hours`. A plural
      rule has nothing to choose on once the value has been rounded to `6.2k`. */
  function tokenPhrase(value) {
    return Str.t("unit.tokens", tokens(value));
  }
  function hourPhrase(value) {
    return Str.t("unit.hours", hours(value));
  }

  /** Join the way the language joins a list: `a, b, c` and `a，b，c`. */
  function list(parts) {
    return parts.join(Str.t("common.listSeparator"));
  }

  var PURPOSES = ["development", "research", "debugging", "conversation", "mixed", "unknown"];

  /** The purpose the reader sees. `unknown` is "other", never "unknown". A key this build
      has never heard of is printed as it came, never dropped. */
  function purpose(key) {
    return PURPOSES.indexOf(key) >= 0 ? Str.t("purpose." + key) : key;
  }

  /** The purpose **inside an observation sentence**. English keeps the engine's own
      label, because that sentence has to equal `app_observation.sentence` word for word
      and the engine writes `labelled unknown` where the interface says "other". Every
      other language gets the reader's word, since there is no English to match. */
  function purposeInSentence(key) {
    return Str.lang() === "en" ? key : purpose(key);
  }

  function weekday(key) {
    return Str.t("weekday." + key);
  }

  global.Fmt = {
    PURPOSES: PURPOSES,
    count: count,
    plain: plain,
    decimal: decimal,
    tokens: tokens,
    hours: hours,
    percent: percent,
    points: points,
    pointsValue: pointsValue,
    stamp: stamp,
    relative: relative,
    day: day,
    shortDay: shortDay,
    writtenOn: writtenOn,
    sessions: sessions,
    commits: commits,
    edits: edits,
    projects: projects,
    tokenPhrase: tokenPhrase,
    hourPhrase: hourPhrase,
    list: list,
    purpose: purpose,
    purposeInSentence: purposeInSentence,
    weekday: weekday,
    fromDay: fromDay,
  };
})(window);
